File: [0494][Medium][Target_Sum]/Target_Sum_2.py
# -*- coding: utf-8 -*-  
# leetcode time     cost : 1688 ms
# leetcode memory   cost : 19.4 MB
# raw DP, slow,
# dp[(i,j)]=dp.get((i−1,j−nums[i]),0)+dp.get((i−1,j+nums[i]),0)
class Solution:
    def findTargetSumWays(self, nums: List[int], S: int) -> int:
        length, dp = len(nums), {(0,0): 1}
        for i in range(1, length+1):
            for j in range(-sum(nums), sum(nums) + 1):
                dp[(i,j)] = dp.get((i - 1, j - nums[i-1]), 0) + dp.get((i - 1, j + nums[i-1]), 0)             
        return dp.get((length, S), 0)

File: [0494][Medium][Target_Sum]/Target_Sum_2.java
// leetcode time     cost : 9 ms
// leetcode memory   cost : 40 MB 
// Time  Complexity: O(N*S)
// Space Complexity: O(N)
// solution 2, DP
public class Solution {
    public int findTargetSumWays(int[] nums, int S) {
        int[] dp = new int[2001];
        dp[nums[0] + 1000] = 1;
        dp[-nums[0] + 1000] += 1;
        for (int i = 1; i < nums.length; i++) {
            int[] next = new int[2001];
            for (int sum = -1000; sum <= 1000; sum++) {
                if (dp[sum + 1000] > 0) {
                    next[sum + nums[i] + 1000] += dp[sum + 1000];
                    next[sum - nums[i] + 1000] += dp[sum + 1000];
                }
            }
            dp = next;
        }
        return S > 1000 ? 0 : dp[S + 1000];
    }
}
